### PR TITLE
[FIX] website_event_track: prevent an error if track duration is too long

### PR DESCRIPTION
**Pull request by mega-odoo**
Before this commit:

- If the track duration is 25-30 hours long then track timeline goes beyond 24 hours which seems absurd. This also leads to duplicate view.

- And if the track duration is too long like 50-100 hours, it leads to a DateTime error at [1].

link [1]: https://github.com/odoo/odoo/blob/20b2116464b5f542a5b89b3fc533e2da2b890944/addons/website_event_track/controllers/event_track.py#L316

Steps to reproduce

- Install a `website_event_track` module.
- Go to Events > Configuration > Event Templates.
- Create a new event template and make a boolean true of Website Submenu', 'Tracks menu Item', 'Track Proposals Menu Item', and 'Register Button'.
- Open 'Events' and create new events.
- Add an event name, 'Date', and also add an event template that we made.
- Click on the 'Tracks' button, Create an Event Tracks, Add a Title, Track Date, and Give a value to a duration field more than 48:00.
- Make this Event Tracks state to 'Published'.
- Click on the 'Go to Website' Button, go into 'Agenda', Error is generated.

After this commit:

Handle tracks with long durations:
- Tracks with durations of 50–100 hours can now be created, even if they extend across multiple days.

Improved time slot allocation:
- The `_split_track_by_days` function now allocates time slots for the next days if a track timeline goes more than 24 hours [2]. This ensures If a track spans more than one day, it should appear in two days separately.

Link 2: https://github.com/odoo/odoo/blob/b2fa69202c6b3ab677973643711ccd561a2edaea/addons/website_event_track/controllers/event_track.py#L307-L315

Clean up unnecessary code.
- e.g. Avoids recalculating `start_date` value in `_get_occupied_cells` which is already available. And reduce iterations over the dictionary when counting locations by days.

Task-3586488

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by adsh-odoo — COMMENTED** [review]
Hello @mega-odoo :slightly_smiling_face: 
A few comments/suggestions.
Thanks!!

**Comment by mega-odoo** [comment]
> Hello @mega-odoo 🙂 A few comments/suggestions. Thanks!!

Hello @adsh-odoo 

The code has been changed according to your suggestions. Thanks :)

File: addons/website_event_track/tests/test_track_internals.py [hunk 1]
```diff
@@ -95,6 +99,48 @@ def test_track_partner_sync(self):
...
+
+            all_timeslot = track_data['time_slots']
+            local_tz = pytz.timezone(self.event_0.date_tz or 'UTC')
+            start_datetime = fields.Datetime.from_string(track_start_day).replace(tzinfo=pytz.utc).astimezone(local_tz)
+            end_datetime = fields.Datetime.from_string(track_end_day).replace(tzinfo=pytz.utc).astimezone(local_tz)
+
+            # compare tracks start datetime
+            self.assertEqual(next(iter(all_timeslot[track_start_day.date()].keys())), start_datetime, 'Track start datetime')
+            # compare tracks end datetime
+            self.assertEqual(list(all_timeslot[track_end_day.date()].keys())[-1], end_datetime, 'Track end datetime')
```
Comment (adsh-odoo) [inline addons/website_event_track/tests/test_track_internals.py:141]:
```suggestion
            self.assertEqual(list(all_timeslot[track_end_day.date()].keys())[-1], end_datetime, 'Track end datetime')
```
IMO It will be better if we check the end_time for both the tracks separately instead of checking the final end_time.

**Review by adsh-odoo — COMMENTED** [review]
Hello @mega-odoo 
Few more comments and then we can proceed.
Thanks!!

**Comment by adsh-odoo** [comment]
And also please don't hesitate to explain the technical things inside the commit message so that it will be easy to understand the issue by reading the commit message.

File: addons/website_event_track/controllers/event_track.py [hunk 2]
```diff
@@ -242,37 +243,37 @@ def _prepare_calendar_values(self, event):
                     'rowspan': duration,  # rowspan
                     'start_date': self._get_locale_time(start_date, lang_code),
                     'end_date': self._get_locale_time(end_date, lang_code),
-                    'occupied_cells': self._get_occupied_cells(track, duration, locations, local_tz)
+                    'occupied_cells': self._get_occupied_cells(track, time_slot, duration, locations, local_tz)
```
[the remark this answers is not in the document]
Comment (mega-odoo) [reply addons/website_event_track/controllers/event_track.py:246]: > Could you please explain why this is necessary?

In the previous code, we retrieved the start timeslot from the tracks and applied time_slot_rounder to it, but we can use the rounded timeslot which is already got from here [1].

link [1]: https://github.com/odoo/odoo/blob/135f94f8f20ef5616f080596934dd5276bd2226d/addons/website_event_track/controllers/event_track.py#L223

**Comment by mega-odoo** [comment]
Hello @adsh-odoo

The code has been changed according to your suggestions. Thanks :)

File: addons/website_event_track/controllers/event_track.py [hunk 3]
```diff
@@ -247,37 +248,37 @@ def _prepare_calendar_values(self, event):
...
+            for time_slot in time_slots:
+                track_day = time_slot.date()
+                if track.location_id not in locations_by_days[track_day]:
+                    locations_by_days[track_day].append(track.location_id)
+
         # split days into 15 minutes time slots
         global_time_slots_by_day = dict((day, {}) for day in days)
         for day, time_slots in time_slots_by_day.items():
             start_time_slot = min(time_slots['start'])
             end_time_slot = max(time_slots['end'])
```
Comment (reth-odoo) [inline addons/website_event_track/controllers/event_track.py]: Can't it be just this? So the loop stops at most at the latest possible slot?
```suggestion
            end_time_slot = min(max(time_slots['end']), next_day)
```

File: addons/website_event_track/tests/test_track_internals.py [hunk 4]
```diff
@@ -178,6 +182,51 @@ def test_track_partner_sync(self):
...
+
+            tracks = self.event_0.track_ids
+            track_start_day = track_end_day = tracks[0].date
+            for track in tracks:
+                track_date = track.date
+                if track_date < track_start_day:
+                    track_start_day = track_date
+                track_end_time = track_date + timedelta(hours=track.duration)
+                if track_end_time > track_end_day:
+                    track_end_day = track_end_time
```
Comment (reth-odoo) [inline addons/website_event_track/tests/test_track_internals.py]: 🛑 Don't generate expected test results with logic.
Just freeze_time and hardcode expected values that we know to be correct without having to think about it.

File: addons/website_event_track/controllers/event_track.py [hunk 5]
```diff
@@ -319,34 +319,33 @@ def _split_track_by_days(self, track, local_tz):
         Also return a set of all the time slots
         """
         start_date = fields.Datetime.from_string(track.date).replace(tzinfo=pytz.utc).astimezone(local_tz)
-        start_datetime = self.time_slot_rounder(start_date, 15)
-        end_datetime = self.time_slot_rounder(start_datetime + timedelta(hours=(track.duration or 0.25)), 15)
-        time_slots_count = int(((end_datetime - start_datetime).total_seconds() / 3600) * 4)
+        track_start_time_slot = start_time_slot = self.time_slot_rounder(start_date, 15)
+        end_datetime = self.time_slot_rounder(start_time_slot + timedelta(hours=(track.duration or 0.25)), 15)
```
Comment (reth-odoo) [inline addons/website_event_track/controllers/event_track.py]: 😴-coder proofing while I'm in the 😴-zone

Let's try keeping the names different enough, and I'd shorten "time_slot" to "slot" since in this context it's implicit
```suggestion
        track_start = current_slot_start = self.time_slot_rounder(start_date, 15)
        current_slot_end = self.time_slot_rounder(current_slot_start + timedelta(hours=(track.duration or 0.25)), 15)
```

File: addons/website_event_track/controllers/event_track.py [hunk 6]
```diff
@@ -319,34 +319,33 @@ def _split_track_by_days(self, track, local_tz):
...
+        time_slots_count = int(((end_datetime - start_time_slot).total_seconds() / 3600) * 4)
```
Comment (reth-odoo) [inline addons/website_event_track/controllers/event_track.py]: Could be me being sleepy but either way, code should be 😴-dev-proof

Let's try keeping the names different enough, and I'd shorten "time_slot" to "slot" since in this context it's implicit
```suggestion
        track_start = day_start = self.time_slot_rounder(start_date, 15)
        track_end = self.time_slot_rounder(track_start + timedelta(hours=(track.duration or 0.25)), 15)
        total_slot_count = int(((end_datetime - start_time_slot).total_seconds() / 3600) * 4)
```
Also `time_slots_by_day_start_time` -> `slot_starts_by_day`

File: addons/website_event_track/controllers/event_track.py [hunk 7]
```diff
@@ -319,34 +319,33 @@ def _split_track_by_days(self, track, local_tz):
...
 
-        time_slots_by_day_start_time = {start_datetime: 0}
+        time_slots_by_day_start_time = {start_time_slot: 0}
+        next_day = (start_time_slot + timedelta(days=1)).date()
         for i in range(0, time_slots_count):
             # If the new time slot is still on the current day
-            next_day = (start_datetime + timedelta(days=1)).date()
-            if (start_datetime + timedelta(minutes=15*i)).date() <= next_day:
-                time_slots_by_day_start_time[start_datetime] += 1
+            time_slots = track_start_time_slot + timedelta(minutes=15 * i)
```
Comment (reth-odoo) [inline addons/website_event_track/controllers/event_track.py]: Naming should make sense, this isn't "slots", it's a datetime.
Also can't we just add 15 minutes to the current slot start now that we have that? Should be marginally faster.
```suggestion
            slot_end = track_start + timedelta(minutes=15)
```

File: addons/website_event_track/tests/test_track_internals.py [hunk 8]
```diff
@@ -179,8 +183,50 @@ def test_track_partner_sync(self):
...
+            local_tz = pytz.timezone(self.event_0.date_tz or 'UTC')
+            expected_start_datetime = local_tz.localize(datetime(2025, 1, 11, 11, 0, 0))
+            expected_end_datetime = local_tz.localize(datetime(2025, 1, 14, 9, 0, 0))
+            # compare tracks start datetime
+            self.assertEqual(next(iter(all_timeslot[track_start_day.date()].keys())), expected_start_datetime, 'Track start datetime')
+            # compare tracks end datetime
+            self.assertEqual(list(all_timeslot[track_end_day.date()].keys())[-1], expected_end_datetime, 'Track end datetime')
+
 
-@tagged('at_install', '-post_install')  # LEGACY at_install
```
Comment (reth-odoo) [inline addons/website_event_track/tests/test_track_internals.py]: Mistake from rebase I presume?

File: addons/website_event_track/tests/test_track_internals.py [hunk 9]
```diff
@@ -179,8 +183,50 @@ def test_track_partner_sync(self):
             new_track.contact_phone, customer.phone,
             'Track customer should take over existing contact phone value')
 
+    def test_prepare_calendar_values(self):
```
Comment (reth-odoo) [inline addons/website_event_track/tests/test_track_internals.py]: 🔤 order
Also docstring is always nice (e.g. "Ensure time slots are generated properly for website agenda.")

File: addons/website_event_track/tests/test_track_internals.py [hunk 10]
```diff
@@ -179,8 +183,50 @@ def test_track_partner_sync(self):
...
+                    track_end_day = track_end_time
+
+            all_timeslot = track_data['time_slots']
+            local_tz = pytz.timezone(self.event_0.date_tz or 'UTC')
+            expected_start_datetime = local_tz.localize(datetime(2025, 1, 11, 11, 0, 0))
+            expected_end_datetime = local_tz.localize(datetime(2025, 1, 14, 9, 0, 0))
+            # compare tracks start datetime
+            self.assertEqual(next(iter(all_timeslot[track_start_day.date()].keys())), expected_start_datetime, 'Track start datetime')
+            # compare tracks end datetime
+            self.assertEqual(list(all_timeslot[track_end_day.date()].keys())[-1], expected_end_datetime, 'Track end datetime')
```
Comment (reth-odoo) [inline addons/website_event_track/tests/test_track_internals.py]: I would just explicitly check the results. You can generate a list/dict with comprehension and addition of lists, based on what we expect to see. Then have simple equality checks to ensure the actual data matches expected data.

This way we have a clear idea of what data we expect, how many intervals, when they start, ...

File: addons/website_event_track/controllers/event_track.py [hunk 11]
```diff
@@ -247,37 +248,37 @@ def _prepare_calendar_values(self, event):
...
+                track_day = time_slot.date()
+                if track.location_id not in locations_by_days[track_day]:
+                    locations_by_days[track_day].append(track.location_id)
+
         # split days into 15 minutes time slots
         global_time_slots_by_day = dict((day, {}) for day in days)
         for day, time_slots in time_slots_by_day.items():
             start_time_slot = min(time_slots['start'])
-            end_time_slot = max(time_slots['end'])
+            next_day = (start_time_slot + timedelta(days=1)).replace(hour=0, minute=0, second=0)
```
Comment (reth-odoo) [inline addons/website_event_track/controllers/event_track.py]: Is replacing hour and minutes sufficient since we're supposed to round to the closest quarter of an hour?

I believe so but if not, would `datetime.combine(start_time_slot.date(), datetime.min())` to ensure we cover milli/micro/nanoseconds and whatnot

File: addons/website_event_track/controllers/event_track.py [hunk 12]
```diff
@@ -247,37 +248,37 @@ def _prepare_calendar_values(self, event):
...
+            end_time_slot = min(max(time_slots['end']), next_day)
 
             time_slots_count = int(((end_time_slot - start_time_slot).total_seconds() / 3600) * 4)
             current_time_slot = start_time_slot
             for _i in range(0, time_slots_count + 1):
-                global_time_slots_by_day[day][current_time_slot] = tracks_by_rounded_times.get(current_time_slot, {})
+                global_time_slots_by_day[day][current_time_slot] = tracks_by_rounded_times.get(current_time_slot, {}) if current_time_slot < next_day else {}
```
Comment (reth-odoo) [inline addons/website_event_track/controllers/event_track.py]: Is this still necessary considering `end_time_slot` shouldn't allow to go over the day?

Comment (mega-odoo) [reply addons/website_event_track/controllers/event_track.py]: > Is this still necessary considering `end_time_slot` shouldn't allow to go over the day?

Yes, to ensure the day span doesn’t go over the current day like this

https://drive.google.com/file/d/1CfxjWay0aWbbYYVyOUIPpPiaJHXcxucj/view?usp=drive_link

File: addons/website_event_track/controllers/event_track.py (hunk 11 above)
Comment (mega-odoo) [reply addons/website_event_track/controllers/event_track.py]: > Is replacing hour and minutes sufficient since we're supposed to round to the closest quarter of an hour?
> 
> I believe so but if not, would `datetime.combine(start_time_slot.date(), datetime.min())` to ensure we cover milli/micro/nanoseconds and whatnot

In Python, datetime.datetime(2025, 11, 14, 0, 0) and datetime.datetime(2025, 11, 14, 0, 0, 0) are the same object.

Python omits the seconds (0) when printing if they are zero

Comment (reth-odoo) [reply addons/website_event_track/controllers/event_track.py]: My point was, why do we set second to 0 but not millisecond here?

File: addons/website_event_track/controllers/event_track.py (hunk 12 above)
Comment (reth-odoo) [reply addons/website_event_track/controllers/event_track.py]: But above we do `end_time_slot = min(max(time_slots['end']), next_day)`, which ends up defining how often this loop runs. So how can we ever be in a case where this condition is False?

Comment (mega-odoo) [reply addons/website_event_track/controllers/event_track.py]: > But above we do `end_time_slot = min(max(time_slots['end']), next_day)`, which ends up defining how often this loop runs. So how can we ever be in a case where this condition is False?

try with this scenario: - `Track date - 11/18/2025 11:00:00 PM, Duration - 2 hour` 

Each hours have 4 time_slots_count, total 8 time_slots_count for 2 hours

You have a dictionary `tracks_by_rounded_times` that only contains two exact datetime keys:

https://drive.google.com/file/d/1ijFwPJmeN_ZySuW-XKxci5rGNW_IbB6z/view?usp=drive_link

2025-11-18 23:00
2025-11-19 00:00

Loop goes through time in 15-minute steps, like:

day 1 - 2025-11-18

23:00  --> we want full data for this hour, like rowspan, start_date, end_date
23:15
23:30
23:45
00:00  --> we want full data for this hour, like rowspan, start_date, end_date, `but not for this day`

day 1 - 2025-11-19

00:00  --> we want full data for this hour, like rowspan, start_date, end_date
00:15
00:30
00:45
01:00

Loop 1 → 23:00
This matches a key in `tracks_by_rounded_times`, `datetime.datetime(2025, 11, 18, 23, 0)`

Loop 2–4 → 23:15, 23:30, 23:45
do NOT exist in dictionary keys

Loop 5 → 00:00 (next day)
This matches the 2nd key in the `tracks_by_rounded_times`, `datetime.datetime(2025, 11, 19, 0, 0)`

As we want this [2] type of key pair value to show the exact event track [3]

------------------------------------------------------------------------------------------------------------------------------------------------------------------
with condition `(if current_time_slot < end_time_slot else {})` data [2], and event track [3]
link [2] - https://drive.google.com/file/d/1s1ZSgCLAk6PPlDUTNxRUIJkatqpV7v6j/view?usp=drive_link

link [3] -https://drive.google.com/file/d/1KrccPn13rdZGJZQNzVEF1pGN0kL8_w-L/view?usp=drive_link

------------------------------------------------------------------------------------------------------------------------------------------------------------------

without condition  data [4], event track with this data [5]
link [4] - https://drive.google.com/file/d/1vytMgIOUYxtQFNcKijo5SoVLnZ0_cTxB/view?usp=drive_link

link [5] -https://drive.google.com/file/d/17bAlAWG5v5hRfNYqxnFAeYmF9gUUaTHp/view?usp=drive_link


```
In short, we have to put the caps at 00:00 so the current day never includes the next day's time slot.
````

Comment (reth-odoo) [reply addons/website_event_track/controllers/event_track.py]: @mega-odoo Can't we just do this then `last_slot_start = start_time_slot + timedelta(days=1).replace(hour=0, minute=0, second=0, microsecond=0) - timedelta(minutes=15)` (probably to multi-line)?

Also can you rebase?
As a note too, prefer pastebin to screenshots of code as that's much easier to work with as the recipient.

Comment (mega-odoo) [reply addons/website_event_track/controllers/event_track.py]: > @mega-odoo Can't we just do this then `last_slot_start = start_time_slot + timedelta(days=1).replace(hour=0, minute=0, second=0, microsecond=0) - timedelta(minutes=15)` (probably to multi-line)?
> 
> Also can you rebase? As a note too, prefer pastebin to screenshots of code as that's much easier to work with as the recipient.

Means we need to define `last_slot_start` to remove this `(if current_time_slot < end_time_slot else {})` condition?

**Review by reth-odoo — APPROVED** [review]
